File: leetcode/dp/fib-style/house-robber.cpp
// https://leetcode.com/problems/house-robber

class Solution {
   int cache[101];
public:
    Solution(){
        fill_n(cache, 101, -1);
    }

    int helper(vector<int>& nums, int n){
        if(n >= nums.size())    return 0;
        if(cache[n]!=-1)    return cache[n];
        
        return cache[n] = max(nums[n] + helper(nums, n+2), helper(nums, n+1));
    }
    
    int rob(vector<int>& nums) {
        return helper(nums, 0);
    }
};

File: leetcode/dp/fib-style/house-robber.java
// https://leetcode.com/problems/house-robber

class Solution {
        int[] dp = new int[101];
    public Solution(){
        for (int i=0;i<101;i++){
            dp[i] = -1;
        }
    }
    public int rob(int[] nums) {
        return r(nums,nums.length-1);

    }
    public int r(int[] nums, int m){
        if (m ==0){
            return nums[0];
        }
        if (m==1){
            return Math.max(nums[0],nums[1]);
        }
        if (dp[m] != -1){
            return dp[m];
        }
        dp[m] = Math.max(r(nums,m-1), r(nums,m-2) + nums[m]);
        return dp[m];
    }
}